importccl: support IMPORT in mixed-version cluster

pbardea opened the pull request:
This commit adds support for running IMPORT in a mixed-version cluster.
Note, that it does not add support for running an IMPORT during the
upgrade of a node however. The IMPORT job would need to be restarted in
that case.

Release note (sql change): Add support for running IMPORT in a
mixed-version cluster.

pbardea committed:
importccl: support IMPORT in mixed-version cluster

This commit adds support for running IMPORT in a mixed-version cluster.
Note, that it does not add support for running an IMPORT during the
upgrade of a node however. The IMPORT job would need to be restarted in
that case.

Release note (sql change): Add support for running IMPORT in a
mixed-version cluster.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/57382)
<!-- Reviewable:end -->

dt commented on pkg/sql/distsql_plan_bulk.go:
```diff
@@ -52,3 +52,15 @@ func (dsp *DistSQLPlanner) SetupAllNodesPlanning(
+
+// GetAllCompatibleNodes returns all nodes that are OK to use in the DistSQL
+// plan.
+func GetAllCompatibleNodes(planCtx *PlanningCtx) []roachpb.NodeID {
```
should we just have SetupAllNodesPlanning do this ?

pbardea replied:
yep

pbardea commented:
I believe that this also fixes https://github.com/cockroachdb/cockroach/issues/57503.

Will add tests.

pbardea committed:
roachtest: add import/mixed-version roachtest

Before supporting mixed-versions runs of IMPORT this test would fail
with:
```
importing fixture: importing table history: pq: version mismatch in flow request: 42; this node accepts 35 through 36
```

It now succeeds. Note that this test exercises that IMPORT can run in a
mixed-version cluster, but not during an upgrade of nodes in the
cluster.

Release note: None

pbardea committed:
roachtest: add import/decomissed roachtest

This commit adds a roachtest that runs an IMPORT on a cluster with a
decommissioned node.

Release note: None

pbardea commented:
Updated with some roachtests that exercise running import in mixed-version clusters and clusters with a decommissioned node.
RFAL

pbardea commented:
Updated with some roachtests that exercise running import in mixed-version clusters and clusters with a decommissioned node.
RFAL

dt commented on pkg/ccl/backupccl/backup_processor_planning.go:
```diff
@@ -48,6 +48,8 @@ func distBackup(
 	dsp := execCtx.DistSQLPlanner()
 	evalCtx := execCtx.ExtendedEvalContext()
 
+	// We don't return the compatible nodes here since PartitionSpans will
```
not a change for this PR but now I'm sorta wondering why we even SetupAllNodesPlanning if we're not planning on _all_ nodes

pbardea replied:
we still need SetupAllNodesPlanning here so that the planCtx can learn about the node status of all of the nodes. I think there's an argument to split out `getAllPlannableNodes` out again, which is just a pure function on the `planCtx`. We only need to `getAllPlannableNodes` in import and restore where we round-robin distribute work.

Even in backup, we still need `planCtx` to know about the status of all nodes so it only plans on healthy ones.

dt replied:
Makes sense, thanks!

pbardea commented:
TFTR!
bors r=dt

pbardea commented:
It doesn't look like this needs backporting to versions before 20.2. It looks like https://github.com/cockroachdb/cockroach/blob/release-20.1/pkg/sql/distsql_physical_planner.go#L1064 never populates the NodeAddresses map for unhealthy nodes so https://github.com/cockroachdb/cockroach/blob/release-20.1/pkg/sql/distsql_plan_csv.go#L123 just did the right thing before
